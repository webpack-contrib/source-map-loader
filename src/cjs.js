const loader = require("./index");

module.exports = loader;
module.exports.raw = loader.raw;
