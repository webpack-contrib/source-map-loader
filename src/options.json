{
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "filterSourceMappingUrl": {
      "instanceof": "Function"
    }
  }
}
