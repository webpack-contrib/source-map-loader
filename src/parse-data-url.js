import MIMEType from "whatwg-mimetype";
import { atob } from "abab";

function isASCIIHex(c) {
  return (
    (c >= 0x30 && c <= 0x39) ||
    (c >= 0x41 && c <= 0x46) ||
    (c >= 0x61 && c <= 0x66)
  );
}

function percentDecodeBytes(input) {
  const output = new Uint8Array(input.byteLength);
  let outputIndex = 0;

  for (let i = 0; i < input.byteLength; ++i) {
    const byte = input[i];

    if (byte !== 0x25) {
      output[outputIndex] = byte;
    } else if (
      byte === 0x25 &&
      (!isASCIIHex(input[i + 1]) || !isASCIIHex(input[i + 2]))
    ) {
      output[outputIndex] = byte;
    } else {
      output[outputIndex] = parseInt(
        String.fromCodePoint(input[i + 1], input[i + 2]),
        16
      );
      i += 2;
    }

    outputIndex += 1;
  }

  return output.slice(0, outputIndex);
}

export default function parseDataUrl(stringInput) {
  let parsedUrl;

  try {
    parsedUrl = new URL(stringInput);
  } catch (error) {
    return null;
  }

  if (parsedUrl.protocol !== "data:") {
    return null;
  }

  parsedUrl.hash = "";

  // `5` is value of `'data:'.length`
  const input = parsedUrl.toString().substring(5);

  let position = 0;
  let mimeType = "";

  while (position < input.length && input[position] !== ",") {
    mimeType += input[position];
    position += 1;
  }

  mimeType = mimeType.replace(/^[ \t\n\f\r]+/, "").replace(/[ \t\n\f\r]+$/, "");

  if (position === input.length) {
    return null;
  }

  position += 1;

  const encodedBody = input.substring(position);

  let body = Buffer.from(percentDecodeBytes(Buffer.from(encodedBody, "utf-8")));

  // Can't use /i regexp flag because it isn't restricted to ASCII.
  const mimeTypeBase64MatchResult = /(.*); *[Bb][Aa][Ss][Ee]64$/.exec(mimeType);

  if (mimeTypeBase64MatchResult) {
    const stringBody = body.toString("binary");
    const asString = atob(stringBody);

    if (asString === null) {
      return null;
    }

    body = Buffer.from(asString, "binary");

    [, mimeType] = mimeTypeBase64MatchResult;
  }

  if (mimeType.startsWith(";")) {
    mimeType = `text/plain ${mimeType}`;
  }

  let mimeTypeRecord;

  try {
    mimeTypeRecord = new MIMEType(mimeType);
  } catch (e) {
    mimeTypeRecord = new MIMEType("text/plain;charset=US-ASCII");
  }

  return { mimeType: mimeTypeRecord, body };
}
