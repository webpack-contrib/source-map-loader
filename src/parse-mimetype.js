const removeLeadingAndTrailingHTTPWhitespace = (string) =>
  string.replace(/^[ \t\n\r]+/, "").replace(/[ \t\n\r]+$/, "");

const removeTrailingHTTPWhitespace = (string) =>
  string.replace(/[ \t\n\r]+$/, "");

const isHTTPWhitespaceChar = (char) =>
  char === " " || char === "\t" || char === "\n" || char === "\r";

const solelyContainsHTTPTokenCodePoints = (string) =>
  /^[-!#$%&'*+.^_`|~A-Za-z0-9]*$/.test(string);

const soleyContainsHTTPQuotedStringTokenCodePoints = (string) =>
  /^[\t\u0020-\u007E\u0080-\u00FF]*$/.test(string);

const asciiLowercase = (string) =>
  string.replace(/[A-Z]/g, (l) => l.toLowerCase());

function parcer(data) {
  const input = removeLeadingAndTrailingHTTPWhitespace(data);

  let position = 0;
  let type = "";
  while (position < input.length && input[position] !== "/") {
    type += input[position];
    position += 1;
  }

  if (type.length === 0 || !solelyContainsHTTPTokenCodePoints(type)) {
    return null;
  }

  if (position >= input.length) {
    return null;
  }

  // Skips past "/"
  position += 1;

  let subtype = "";
  while (position < input.length && input[position] !== ";") {
    subtype += input[position];
    position += 1;
  }

  subtype = removeTrailingHTTPWhitespace(subtype);

  if (subtype.length === 0 || !solelyContainsHTTPTokenCodePoints(subtype)) {
    return null;
  }

  const mimeType = {
    parameters: new Map(),
  };

  while (position < input.length) {
    // Skip past ";"
    position += 1;

    while (isHTTPWhitespaceChar(input[position])) {
      position += 1;
    }

    let parameterName = "";
    while (
      position < input.length &&
      input[position] !== ";" &&
      input[position] !== "="
    ) {
      parameterName += input[position];
      position += 1;
    }
    parameterName = asciiLowercase(parameterName);

    if (position < input.length) {
      if (input[position] === ";") {
        // eslint-disable-next-line no-continue
        continue;
      }

      // Skip past "="
      position += 1;
    }

    let parameterValue = "";
    if (input[position] === '"') {
      position += 1;

      // eslint-disable-next-line no-constant-condition
      while (true) {
        while (
          position < input.length &&
          input[position] !== '"' &&
          input[position] !== "\\"
        ) {
          parameterValue += input[position];
          position += 1;
        }

        break;
      }

      while (position < input.length && input[position] !== ";") {
        position += 1;
      }
    } else {
      while (position < input.length && input[position] !== ";") {
        parameterValue += input[position];
        position += 1;
      }

      parameterValue = removeTrailingHTTPWhitespace(parameterValue);

      if (parameterValue === "") {
        // eslint-disable-next-line no-continue
        continue;
      }
    }

    if (
      parameterName.length > 0 &&
      solelyContainsHTTPTokenCodePoints(parameterName) &&
      soleyContainsHTTPQuotedStringTokenCodePoints(parameterValue) &&
      !mimeType.parameters.has(parameterName)
    ) {
      mimeType.parameters.set(parameterName, parameterValue);
    }
  }

  return mimeType;
}

function getMimeCharset(input) {
  const result = parcer(input);

  return result.parameters.get("charset");
}

export default getMimeCharset;
