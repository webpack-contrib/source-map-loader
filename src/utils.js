import path from "path";
import urlUtils from "url";

import sourceMap from "source-map";
import { decode } from "iconv-lite";

import parseDataURL from "./parse-data-url";
import labelsToNames from "./labels-to-names";

// Matches only the last occurrence of sourceMappingURL
const innerRegex = /\s*[#@]\s*sourceMappingURL\s*=\s*([^\s'"]*)\s*/;

/* eslint-disable prefer-template */
const sourceMappingURLRegex = RegExp(
  "(?:" +
    "/\\*" +
    "(?:\\s*\r?\n(?://)?)?" +
    "(?:" +
    innerRegex.source +
    ")" +
    "\\s*" +
    "\\*/" +
    "|" +
    "//(?:" +
    innerRegex.source +
    ")" +
    ")" +
    "\\s*"
);
/* eslint-enable prefer-template */

function labelToName(label) {
  const labelLowercase = String(label).trim().toLowerCase();

  return labelsToNames[labelLowercase] || null;
}

async function flattenSourceMap(map) {
  const consumer = await new sourceMap.SourceMapConsumer(map);
  const generatedMap = map.file
    ? new sourceMap.SourceMapGenerator({
        file: map.file,
      })
    : new sourceMap.SourceMapGenerator();

  consumer.sources.forEach((sourceFile) => {
    const sourceContent = consumer.sourceContentFor(sourceFile, true);
    generatedMap.setSourceContent(sourceFile, sourceContent);
  });

  consumer.eachMapping((mapping) => {
    const { source } = consumer.originalPositionFor({
      line: mapping.generatedLine,
      column: mapping.generatedColumn,
    });

    const mappings = {
      source,
      original: {
        line: mapping.originalLine,
        column: mapping.originalColumn,
      },
      generated: {
        line: mapping.generatedLine,
        column: mapping.generatedColumn,
      },
    };

    if (source) {
      generatedMap.addMapping(mappings);
    }
  });

  return generatedMap.toJSON();
}

function getSourceMappingURL(code) {
  const lines = code.split(/^/m);
  let match;

  for (let i = lines.length - 1; i >= 0; i--) {
    match = lines[i].match(sourceMappingURLRegex);
    if (match) {
      break;
    }
  }

  const sourceMappingURL = match ? match[1] || match[2] || "" : null;

  return {
    sourceMappingURL: sourceMappingURL
      ? decodeURI(sourceMappingURL)
      : sourceMappingURL,
    replacementString: match ? match[0] : null,
  };
}

function getAbsolutePath(context, request, sourceRoot) {
  if (sourceRoot) {
    if (path.isAbsolute(sourceRoot)) {
      return path.join(sourceRoot, request);
    }

    return path.join(context, sourceRoot, request);
  }

  return path.join(context, request);
}

function fetchFromDataURL(loaderContext, sourceURL) {
  const dataURL = parseDataURL(sourceURL);

  if (dataURL) {
    // https://tools.ietf.org/html/rfc4627
    // JSON text SHALL be encoded in Unicode. The default encoding is UTF-8.
    const encodingName =
      labelToName(dataURL.parameters.get("charset")) || "UTF-8";

    return decode(dataURL.body, encodingName);
  }

  throw new Error(`Failed to parse source map from "data" URL: ${sourceURL}`);
}

async function fetchFromFilesystem(loaderContext, sourceURL) {
  let buffer;

  try {
    buffer = await new Promise((resolve, reject) => {
      loaderContext.fs.readFile(sourceURL, (error, data) => {
        if (error) {
          return reject(error);
        }

        return resolve(data);
      });
    });
  } catch (error) {
    throw new Error(
      `Failed to parse source map from '${sourceURL}' file: ${error}`
    );
  }

  return buffer.toString();
}

async function fetchPathsFromFilesystem(
  loaderContext,
  possibleRequests,
  errorsAccumulator = ""
) {
  let result;

  try {
    result = await fetchFromFilesystem(
      loaderContext,
      possibleRequests[0],
      errorsAccumulator
    );
  } catch (error) {
    // eslint-disable-next-line no-param-reassign
    errorsAccumulator += `${error.message}\n\n`;

    const [, ...tailPossibleRequests] = possibleRequests;

    if (tailPossibleRequests.length === 0) {
      error.message = errorsAccumulator;

      throw error;
    }

    return fetchPathsFromFilesystem(
      loaderContext,
      tailPossibleRequests,
      errorsAccumulator
    );
  }

  return result;
}

async function fetchFromURL(
  loaderContext,
  context,
  url,
  sourceRoot,
  skipReading = false
) {
  // 1. It's an absolute url and it is not `windows` path like `C:\dir\file`
  if (/^[a-z][a-z0-9+.-]*:/i.test(url) && !path.win32.isAbsolute(url)) {
    const { protocol } = urlUtils.parse(url);

    if (protocol === "data:") {
      if (skipReading) {
        return { sourceURL: "" };
      }

      const sourceContent = fetchFromDataURL(loaderContext, url);

      return { sourceURL: "", sourceContent };
    }

    if (skipReading) {
      return { sourceURL: url };
    }

    if (protocol === "file:") {
      const pathFromURL = urlUtils.fileURLToPath(url);
      const sourceURL = path.normalize(pathFromURL);
      const sourceContent = await fetchFromFilesystem(loaderContext, sourceURL);

      return { sourceURL, sourceContent };
    }

    throw new Error(
      `Failed to parse source map: '${url}' URL is not supported`
    );
  }

  // 2. It's a scheme-relative
  if (/^\/\//.test(url)) {
    throw new Error(
      `Failed to parse source map: '${url}' URL is not supported`
    );
  }

  // 3. Absolute path
  if (path.isAbsolute(url)) {
    const sourceURL = path.normalize(url);

    let sourceContent;

    if (!skipReading) {
      const possibleRequests = [sourceURL];

      if (url.startsWith("/")) {
        possibleRequests.push(
          getAbsolutePath(context, sourceURL.slice(1), sourceRoot)
        );
      }

      sourceContent = await fetchPathsFromFilesystem(
        loaderContext,
        possibleRequests
      );
    }

    return { sourceURL, sourceContent };
  }

  // 4. Relative path
  const sourceURL = getAbsolutePath(context, url, sourceRoot);

  let sourceContent;

  if (!skipReading) {
    sourceContent = await fetchFromFilesystem(loaderContext, sourceURL);
  }

  return { sourceURL, sourceContent };
}

export { getSourceMappingURL, fetchFromURL, flattenSourceMap };
