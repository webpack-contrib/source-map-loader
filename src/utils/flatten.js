import sourceMap from 'source-map';

async function FlattenSourceMap(map) {
  const consumer = await new sourceMap.SourceMapConsumer(map);
  let generatedMap;

  if (map.file) {
    generatedMap = new sourceMap.SourceMapGenerator({
      file: map.file,
    });
  } else {
    generatedMap = new sourceMap.SourceMapGenerator();
  }

  consumer.sources.forEach((sourceFile) => {
    const sourceContent = consumer.sourceContentFor(sourceFile, true);
    generatedMap.setSourceContent(sourceFile, sourceContent);
  });

  consumer.eachMapping((mapping) => {
    const { source } = consumer.originalPositionFor({
      line: mapping.generatedLine,
      column: mapping.generatedColumn,
    });

    const mappings = {
      source,
      original: {
        line: mapping.originalLine,
        column: mapping.originalColumn,
      },
      generated: {
        line: mapping.generatedLine,
        column: mapping.generatedColumn,
      },
    };

    if (source) {
      generatedMap.addMapping(mappings);
    }
  });

  return generatedMap.toJSON();
}

module.exports = FlattenSourceMap;
