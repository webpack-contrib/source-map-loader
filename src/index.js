/*
  MIT License http://www.opensource.org/licenses/mit-license.php
  Author Tobias Koppers @sokra
*/
import path from "path";

import schema from "./options.json";
import { getSourceMappingURL, fetchFromURL, flattenSourceMap } from "./utils";

export default async function loader(input, inputMap) {
  const options = this.getOptions(schema);
  const { sourceMappingURL, replacementString } = getSourceMappingURL(input);
  const callback = this.async();

  if (!sourceMappingURL) {
    callback(null, input, inputMap);

    return;
  }

  let behaviourSourceMappingUrl;

  try {
    behaviourSourceMappingUrl =
      typeof options.filterSourceMappingUrl !== "undefined"
        ? options.filterSourceMappingUrl(sourceMappingURL, this.resourcePath)
        : "consume";
  } catch (error) {
    callback(error);

    return;
  }

  // eslint-disable-next-line default-case
  switch (behaviourSourceMappingUrl) {
    case "skip":
      callback(null, input, inputMap);
      return;
    case false:
    case "remove":
      callback(null, input.replace(replacementString, ""), inputMap);
      return;
  }

  let sourceURL;
  let sourceContent;

  try {
    ({ sourceURL, sourceContent } = await fetchFromURL(
      this,
      this.context,
      sourceMappingURL
    ));
  } catch (error) {
    this.emitWarning(error);

    callback(null, input, inputMap);

    return;
  }

  if (sourceURL) {
    this.addDependency(sourceURL);
  }

  let map;

  try {
    map = JSON.parse(sourceContent.replace(/^\)\]\}'/, ""));
  } catch (parseError) {
    this.emitWarning(
      new Error(
        `Failed to parse source map from '${sourceMappingURL}': ${parseError}`
      )
    );

    callback(null, input, inputMap);

    return;
  }

  const context = sourceURL ? path.dirname(sourceURL) : this.context;

  if (map.sections) {
    // eslint-disable-next-line no-param-reassign
    map = await flattenSourceMap(map);
  }

  const resolvedSources = await Promise.all(
    map.sources.map(async (source, i) => {
      // eslint-disable-next-line no-shadow
      let sourceURL;
      // eslint-disable-next-line no-shadow
      let sourceContent;

      const originalSourceContent =
        map.sourcesContent && typeof map.sourcesContent[i] !== "undefined"
          ? map.sourcesContent[i]
          : // eslint-disable-next-line no-undefined
            undefined;
      const skipReading = typeof originalSourceContent !== "undefined";
      let errored = false;

      // We do not skipReading here, because we need absolute paths in sources.
      // This is necessary so that for sourceMaps with the same file structure in sources, name collisions do not occur.
      // https://github.com/webpack-contrib/source-map-loader/issues/51
      try {
        ({ sourceURL, sourceContent } = await fetchFromURL(
          this,
          context,
          source,
          map.sourceRoot,
          skipReading
        ));
      } catch (error) {
        errored = true;

        this.emitWarning(error);
      }

      if (skipReading) {
        sourceContent = originalSourceContent;
      } else if (!errored && sourceURL) {
        this.addDependency(sourceURL);
      }

      // Return original value of `source` when error happens
      return { sourceURL: errored ? source : sourceURL, sourceContent };
    })
  );

  const newMap = { ...map };

  newMap.sources = [];
  newMap.sourcesContent = [];

  delete newMap.sourceRoot;

  resolvedSources.forEach((source) => {
    // eslint-disable-next-line no-shadow
    const { sourceURL, sourceContent } = source;

    newMap.sources.push(sourceURL || "");
    newMap.sourcesContent.push(sourceContent || "");
  });

  const sourcesContentIsEmpty =
    newMap.sourcesContent.filter((entry) => Boolean(entry)).length === 0;

  if (sourcesContentIsEmpty) {
    delete newMap.sourcesContent;
  }

  callback(null, input.replace(replacementString, ""), newMap);
}
