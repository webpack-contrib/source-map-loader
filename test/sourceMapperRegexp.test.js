import { getSourceMappingURL } from "../src/utils";

describe("source-map-loader", () => {
  const cases = [
    "/*#sourceMappingURL=absolute-sourceRoot-source-map.map*/",
    "/*  #sourceMappingURL=absolute-sourceRoot-source-map.map  */",
    "//#sourceMappingURL=absolute-sourceRoot-source-map.map",
    "//@sourceMappingURL=absolute-sourceRoot-source-map.map",
    " //  #sourceMappingURL=absolute-sourceRoot-source-map.map",
    " //  #  sourceMappingURL  =   absolute-sourceRoot-source-map.map  ",
    "// #sourceMappingURL = http://sampledomain.com/external-source-map2.map",
    "// #sourceMappingURL = //sampledomain.com/external-source-map2.map",
    "// @sourceMappingURL=data:application/source-map;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5saW5lLXNvdXJjZS1tYXAuanMiLCJzb3VyY2VzIjpbImlubGluZS1zb3VyY2UtbWFwLnR4dCJdLCJzb3VyY2VzQ29udGVudCI6WyJ3aXRoIFNvdXJjZU1hcCJdLCJtYXBwaW5ncyI6IkFBQUEifQ==",
    `
    with SourceMap

    // #sourceMappingURL = /sample-source-map.map
    // comment
    `,
    `
    with SourceMap
    // #sourceMappingURL = /sample-source-map-1.map
    // #sourceMappingURL = /sample-source-map-2.map
    // #sourceMappingURL = /sample-source-map-last.map
    // comment
    `,
    `"
    /*# sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+" */";`,
    `anInvalidDirective = "\\n/*# sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+" */";`,
    `// # sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+"`,
  ];

  cases.forEach((item) => {
    it(`should work with "${item}" url`, async () => {
      const { sourceMappingURL } = getSourceMappingURL(item);

      expect(sourceMappingURL).toMatchSnapshot("result");
    });
  });
});
