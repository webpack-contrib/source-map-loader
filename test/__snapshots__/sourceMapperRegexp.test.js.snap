// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source-map-loader should work with "
    with SourceMap

    // #sourceMappingURL = /sample-source-map.map
    // comment
    " url: result 1`] = `"/sample-source-map.map"`;

exports[`source-map-loader should work with "
    with SourceMap
    // #sourceMappingURL = /sample-source-map-1.map
    // #sourceMappingURL = /sample-source-map-2.map
    // #sourceMappingURL = /sample-source-map-last.map
    // comment
    " url: result 1`] = `"/sample-source-map-last.map"`;

exports[`source-map-loader should work with " //  #  sourceMappingURL  =   absolute-sourceRoot-source-map.map  " url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with " //  #sourceMappingURL=absolute-sourceRoot-source-map.map" url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with ""
    /*# sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+" */";" url: result 1`] = `null`;

exports[`source-map-loader should work with "/*  #sourceMappingURL=absolute-sourceRoot-source-map.map  */" url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with "/*#sourceMappingURL=absolute-sourceRoot-source-map.map*/" url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with "// # sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+"" url: result 1`] = `"data:application/json;base64,"`;

exports[`source-map-loader should work with "// #sourceMappingURL = //sampledomain.com/external-source-map2.map" url: result 1`] = `"//sampledomain.com/external-source-map2.map"`;

exports[`source-map-loader should work with "// #sourceMappingURL = http://sampledomain.com/external-source-map2.map" url: result 1`] = `"http://sampledomain.com/external-source-map2.map"`;

exports[`source-map-loader should work with "// @sourceMappingURL=data:application/source-map;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5saW5lLXNvdXJjZS1tYXAuanMiLCJzb3VyY2VzIjpbImlubGluZS1zb3VyY2UtbWFwLnR4dCJdLCJzb3VyY2VzQ29udGVudCI6WyJ3aXRoIFNvdXJjZU1hcCJdLCJtYXBwaW5ncyI6IkFBQUEifQ==" url: result 1`] = `"data:application/source-map;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5saW5lLXNvdXJjZS1tYXAuanMiLCJzb3VyY2VzIjpbImlubGluZS1zb3VyY2UtbWFwLnR4dCJdLCJzb3VyY2VzQ29udGVudCI6WyJ3aXRoIFNvdXJjZU1hcCJdLCJtYXBwaW5ncyI6IkFBQUEifQ=="`;

exports[`source-map-loader should work with "//#sourceMappingURL=absolute-sourceRoot-source-map.map" url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with "//@sourceMappingURL=absolute-sourceRoot-source-map.map" url: result 1`] = `"absolute-sourceRoot-source-map.map"`;

exports[`source-map-loader should work with "anInvalidDirective = "\\n/*# sourceMappingURL=data:application/json;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap))))+" */";" url: result 1`] = `null`;
